File: ios/SafeBox/App/UnlockReveal.swift
import SwiftUI

/// The one place the unlock reveal is specified (iteration-2 decisions §1 and
/// §11), value-for-value the same as Android's `object UnlockReveal`.
///
/// Treatment: "zoom-in reveal" (shared-axis Z). The vault fades in while
/// scaling `initialScale → 1.0`; the calculator fades out in place (no scale).
/// Both run concurrently over `durationMs` on the emphasized-decelerate curve.
///
/// It plays ONLY when the lock state becomes `.unlocked` — from `.locked` or
/// from first-run setup, which deliberately uses the same reveal. Every other
/// change is an instant cut: manual lock, background lock, setup-phase
/// changes, erase-everything, and every `calculatorEpoch` bump (a recreated
/// calculator must never read as a transition).
///
/// Reduced motion (`accessibilityReduceMotion`) drops the scale and keeps an
/// opacity-only crossfade with the same duration and curve.
enum UnlockReveal {
    /// Shared duration (Android `UNLOCK_REVEAL_DURATION_MS`).
    static let durationMs = 260
    static var duration: TimeInterval { TimeInterval(durationMs) / 1000 }

    /// Scale the vault starts from in the full reveal.
    static let initialScale: CGFloat = 0.92

    /// cubic-bezier(0.05, 0.7, 0.1, 1.0) — Material "emphasized decelerate".
    static let curveX1: Double = 0.05
    static let curveY1: Double = 0.7
    static let curveX2: Double = 0.1
    static let curveY2: Double = 1.0

    /// How one lock-state change is put on screen.
    enum Kind: Equatable, Sendable {
        /// Vault opacity 0→1 and scale `initialScale`→1; calculator opacity 1→0.
        case zoomReveal
        /// Reduced-motion fallback: same timing, opacity only, no scale.
        case crossfade
        /// Instant — no transition of any kind.
        case cut

        var animates: Bool { self != .cut }
    }

    /// The pure decision the root view follows: only a change INTO
    /// `.unlocked` animates; the origin may be `.locked` or any
    /// `.firstRunSetup` phase. Same-state re-renders (epoch bumps) are cuts.
    static func kind(from: LockState, to: LockState, reduceMotion: Bool) -> Kind {
        guard to == .unlocked, from != .unlocked else { return .cut }
        return reduceMotion ? .crossfade : .zoomReveal
    }

    /// The scale the vault starts from for `kind`, or nil when it must not
    /// scale (reduced motion, or no transition at all).
    static func vaultInitialScale(for kind: Kind) -> CGFloat? {
        kind == .zoomReveal ? initialScale : nil
    }

    /// `durationMs` on the shared curve. Used as the ONLY animation the root
    /// view ever attaches to a lock-state change.
    static var animation: Animation {
        .timingCurve(curveX1, curveY1, curveX2, curveY2, duration: duration)
    }

    /// Transition for the vault surface. Only insertion is ever animated —
    /// removal (lock) is a cut and additionally runs with animations disabled.
    static func vaultTransition(for kind: Kind) -> AnyTransition {
        guard kind.animates else { return .identity }
        let fade = AnyTransition.opacity
        let insertion = vaultInitialScale(for: kind).map { fade.combined(with: .scale(scale: $0)) } ?? fade
        return .asymmetric(insertion: insertion, removal: .identity)
    }

    /// Transition for the calculator surface: it fades out in place and never
    /// scales. Insertion (lock) is a cut. Constant on purpose — a removal
    /// transition is captured from the last frame the view was on screen, so
    /// it must not depend on the state being transitioned to.
    static var calculatorTransition: AnyTransition {
        .asymmetric(insertion: .identity, removal: .opacity)
    }
}

File: android/app/src/main/java/com/calcplus/calculator/app/UnlockReveal.kt
package com.calcplus.calculator.app

import android.content.Context
import android.provider.Settings
import androidx.compose.animation.ContentTransform
import androidx.compose.animation.EnterTransition
import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.Easing
import androidx.compose.animation.core.FiniteAnimationSpec
import androidx.compose.animation.core.snap
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import com.calcplus.calculator.core.lock.LockState

/**
 * The calculator → vault "zoom-in reveal" (iteration-2 decisions §1): the one
 * place that owns the reveal's constants and the rule deciding when it plays.
 * Mirrors iOS `enum UnlockReveal`; the numbers are shared verbatim.
 *
 * The rule is deliberately narrow. A change INTO [LockState.Unlocked] — from
 * [LockState.Locked] or from first-run setup ([LockState.NeedsSetup]) — is the
 * only lock-state change that animates. Locking in any form (manual, background,
 * erase), setup ↔ locked, and every `calculatorEpoch` bump are instant cuts:
 * the calculator's recreation must never read as a transition.
 *
 * Android has no reduce-motion flag. Its only signal is the developer option
 * "Remove animations", i.e. `Settings.Global.ANIMATOR_DURATION_SCALE == 0`,
 * which means *no animation at all* — so the reveal becomes a cut there
 * (iOS, which does have `accessibilityReduceMotion`, drops only the scale).
 */
object UnlockReveal {
    /** Reveal length at the default animator scale. Shared with iOS (`UnlockReveal.durationMs`). */
    const val UNLOCK_REVEAL_DURATION_MS = 260

    /** The vault starts at this scale and settles at 1.0; the calculator only fades. */
    const val INITIAL_SCALE = 0.92f

    /** Emphasized-decelerate, cubic-bezier(0.05, 0.7, 0.1, 1.0). Shared with iOS. */
    val EASING: Easing = CubicBezierEasing(0.05f, 0.7f, 0.1f, 1.0f)

    /**
     * Z order of a surface brought in by a [CUT][Kind.CUT] — always a locked
     * surface (calculator or first-run guide), or a vault entering with
     * animations off. Strictly above [REVEAL_Z_INDEX] so that, on a lock, the
     * calculator is drawn OVER the vault from the very first frame: the
     * outgoing vault stays composed for one frame before it is disposed, and
     * not one of its pixels may show through.
     */
    const val CUT_Z_INDEX = 1f

    /**
     * Z order of the vault brought in by a [REVEAL][Kind.REVEAL]. Below the
     * calculator it replaces, which is the depth the motion describes: the
     * calculator fades away in front, the vault grows up from behind it.
     */
    const val REVEAL_Z_INDEX = 0f

    /** What a lock-state change looks like on screen. */
    enum class Kind {
        /** Vault fades in while scaling [INITIAL_SCALE] → 1.0; calculator fades out in place. */
        REVEAL,

        /** Instant switch; nothing animates. */
        CUT,
    }

    /**
     * Pure decision. [REVEAL][Kind.REVEAL] only when [to] is [LockState.Unlocked],
     * [from] is not, and system animations are on ([animatorDurationScale] > 0).
     * Every other combination — including a same-state "change", which is what
     * an epoch bump looks like at this level — is a [CUT][Kind.CUT].
     */
    fun kind(from: LockState, to: LockState, animatorDurationScale: Float): Kind =
        if (to == LockState.Unlocked && from != LockState.Unlocked && animatorDurationScale > 0f) {
            Kind.REVEAL
        } else {
            Kind.CUT
        }

    /**
     * The system "Animator duration scale". `0` is the "Remove animations"
     * developer option; an unset value reads as the platform default `1`.
     * Read at each lock-state change (not cached) so toggling the option while
     * the app runs is honored without a restart.
     */
    fun animatorDurationScale(context: Context): Float =
        Settings.Global.getFloat(
            context.contentResolver,
            Settings.Global.ANIMATOR_DURATION_SCALE,
            1f,
        )

    /**
     * The `AnimatedContent` transform for a [kind]. Never animates size
     * (`sizeTransform = null`): both surfaces fill the window, and a size
     * animation would read as a glitch.
     *
     * The CUT does not rely on `ExitTransition.None`. `AnimatedContent` keeps
     * tracking an in-flight enter (the vault's fade+scale) when the target
     * flips back mid-reveal, and with no exit of its own the outgoing surface
     * would then *spring* from its partial alpha/scale to fully opaque over
     * ~0.5 s before being disposed. Snapping alpha to 0 (and scale to 1, so
     * the tracked scale-in has an explicit zero-length target too) gives the
     * exit a zero-length target, which is what pins the FAST DISPOSAL of the
     * outgoing surface: on a plain lock it is gone on the next frame.
     *
     * The snap is NOT what guarantees that no vault pixel shows. When a lock
     * interrupts a running reveal, animation-core's `Transition` replaces a
     * non-spring spec on an in-flight animation with its interruption spring,
     * so the vault's alpha eases to 0 over ~150–300 ms instead of snapping,
     * and the reused calculator springs from its mid-fade alpha back to 1
     * (~400 ms) — the vault would be visible THROUGH the translucent
     * calculator for those frames. That case is closed by the draw guard in
     * `SafeBoxApp` (`drawWithContent` skips an exiting Unlocked surface),
     * which does not depend on the alpha animation at all. Together: the snap
     * keeps disposal quick, the draw guard keeps the vault invisible.
     */
    fun contentTransform(kind: Kind): ContentTransform = when (kind) {
        Kind.REVEAL -> ContentTransform(
            targetContentEnter = fadeIn(spec()) + scaleIn(spec(), initialScale = INITIAL_SCALE),
            initialContentExit = fadeOut(spec()),
            targetContentZIndex = REVEAL_Z_INDEX,
            sizeTransform = null,
        )
        Kind.CUT -> ContentTransform(
            targetContentEnter = EnterTransition.None,
            initialContentExit = fadeOut(cutSpec()) + scaleOut(cutSpec(), targetScale = 1f),
            targetContentZIndex = CUT_Z_INDEX,
            sizeTransform = null,
        )
    }

    private fun spec(): FiniteAnimationSpec<Float> =
        tween(durationMillis = UNLOCK_REVEAL_DURATION_MS, easing = EASING)

    /** Zero-duration: the value is the target from the first animated frame. */
    private fun cutSpec(): FiniteAnimationSpec<Float> = snap()
}
